### PR TITLE
proxy not providing user info should cause error

### LINKED ISSUE #42437: auth proxies can't proxy anonymous requests
When an auth proxy tries to proxy an anonymous request, the user.Info ends up with the group system:authenticated, which is wrong.

### PR DESCRIPTION
**Pull request by deads2k**
Fixes #42437

When using a front proxy authenticator it should behave like this:
 1. unrecognized client-cert: skip
 1. unrecognized subject: skip
 1. recognized client-cert/subject, missing username: fail (the data provided was invalid)

In addition, the group information should come from the front proxy and not be later manipulated.  Having that separation eliminates weird cases like the proxy saying a user is `system:anonymous` and the authentication chain adding group `system:authenticated`.

@kubernetes/sig-auth-pr-reviews 

```release-note
You cannot use a front proxy client certificate as front proxy client and a direct user against the API.
```

### CODE REVIEW

**Comment by deads2k** [comment]
@ncdc @enj ptal.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/delegating.go [hunk 1]
```diff
@@ -111,5 +95,23 @@ func (c DelegatingAuthenticatorConfig) New() (authenticator.Request, *spec.Secur
 	if c.Anonymous {
 		authenticator = unionauth.NewFailOnError(authenticator, anonymous.NewAuthenticator())
 	}
+
+	// front proxies go first, but since they provide group information, we need to ensure that we don't add *more*
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/delegating.go]: watch the `if len(authenticators) == 0` case above

Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/delegating.go]: might be more straightforward to build the top-level chain at the end, e.g.

```
topLevelAuthenticators []authenticator.Request{}
if c.RequestHeaderConfig != nil {
  topLevelAuthenticators = append(topLevelAuthenticators, ...)
}

if len(authenticators) > 0 {
  topLevelAuthenticators = append(topLevelAuthenticators, group.NewGroupAdder(unionauth.New(authenticators...), []string{user.AllAuthenticated}))
}

if c.Anonymous {
  topLevelAuthenticators = append(topLevelAuthenticators, anonymous.NewAuthenticator())
}

return unionauth.NewFailOnError(topLevelAuthenticators...)
```

**Comment by deads2k** [comment]
> watch the if len(authenticators) == 0 case above

made it a switch case covering all four cases.

File: staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go [hunk 2]
```diff
@@ -122,10 +123,13 @@ func (a *Verifier) AuthenticateRequest(req *http.Request) (user.Info, bool, erro
 	}
 
 	if _, err := req.TLS.PeerCertificates[0].Verify(optsCopy); err != nil {
-		return nil, false, err
+		if invalidCertErr, ok := err.(x509.CertificateInvalidError); ok && invalidCertErr.Reason == x509.Expired {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go]: unknown authority is a separate error (x509.UnknownAuthorityError). all the CertificateInvalidError reasons seem like things we'd want to bubble on.

Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go]: done

File: staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader.go [hunk 3]
```diff
@@ -107,7 +108,7 @@ func NewSecure(clientCA string, proxyClientNames []string, nameHeaders []string,
 func (a *requestHeaderAuthRequestHandler) AuthenticateRequest(req *http.Request) (user.Info, bool, error) {
 	name := headerValue(req.Header, a.nameHeaders)
 	if len(name) == 0 {
-		return nil, false, nil
+		return nil, false, errors.New("proxy did not provide user information")
 	}
 	groups := allHeaderValues(req.Header, a.groupHeaders)
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader.go]: Don't we need to require at least one group as well?

Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader.go]: nope

Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader.go]: hmm, bears more thought. will discuss tomorrow.

**Comment by deads2k** [comment]
We could make an empty usrername from a front proxy turn into system:anonymous and system:unauthenticated and then have a conditional group adder for system:authenticated.  It would mean I went through a lot of motion today, but might match expectation.

**Comment by enj** [comment]
How about:

1. `username` and `groups` required, if missing, error
2. if `username` == `system:anonymous`, requires `groups` == `[system:unauthenticated]`, `extras` == empty, else error
3. if any other `username`, `groups` must contain `system:authenticated` and must not contain `system:unauthenticated`, else error

I see a few problems with it though:

1. Lots of special casing around `username` and `groups`
2. Not sure if you could go through the front proxy with a token that the API server would understand but the proxy does not (i.e. proxy says you are `system:anonymous` but if you hit the API server directly you would be authenticated as a valid user).

**Comment by deads2k** [comment]
Updated to special case the anonymous user in our proxy.

**Comment by liggitt** [comment]
This has implications for auth proxies that pass through requests with authorization headers without adding user info. Will revisit the proxy client cert fallback issue in 1.7. Go ahead and open a smaller PR for 1.6 that just includes the improvement to the group adder

**Comment by enj** [comment]
Keep open for now I suppose?

**Pull request closed without merging**